File: Leetcode/src/com/leetcode/bfsANDdfs/_994_RottingOranges.java
package com.leetcode.bfsANDdfs;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @Date: 05/02/2020
 * @Description: BFS,
 **/
public class _994_RottingOranges {
    // 直接修改其值，不用visited数组
    public int orangesRotting(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0] == null || grid[0].length == 0) {
            return 0;
        }
        int fresh = 0;
        int m = grid.length, n = grid[0].length;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) {
                    queue.offer(new int[]{i, j});
                } else if (grid[i][j] == 1) {
                    fresh++;
                }
            }
        }
        if (fresh == 0) return 0;
        int res = 0;
        int[][] dirs = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] curr = queue.poll();
                for (int[] dir : dirs) {
                    int x = curr[0] + dir[0], y = curr[1] + dir[1];
                    if (x < 0 || x >= grid.length || y  < 0 || y >= grid[0].length) continue;
                    if (grid[x][y] != 1) continue;
                    fresh--;
                    queue.offer(new int[]{x, y});
                    grid[x][y] = 2;
                }
            }
            res++;
        }
        return fresh > 0 ? -1 : res - 1;
    }
}

File: Leetcode/src/com/leetcode/bfsANDdfs/_994_Rotting_Oranges.py
'''
Date: 10/29/2021 09:49:40
LastEditTime: 10/29/2021 10:25:38
Description: You need to fill out
'''
from queue import Queue
from typing import List


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # BFS
        # 1、 一开始没有注意到使用fresh来统计
        # 2、没有每层遍历（qsize())

        m, n = len(grid), len(grid[0])
        q = Queue()
        fresh = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    # Add rotten orange and expand them.
                    q.put((i, j))
                elif grid[i][j] == 1:
                    fresh += 1
        if fresh == 0:
            return 0
        steps = -1
        dirs = [[-1, 0], [0, -1], [1, 0], [0, 1]]
        while not q.empty():
            for _ in range(q.qsize()):
                x, y = q.get()
                for d in dirs:
                    new_x = x + d[0]
                    new_y = y + d[1]

                    if new_x < 0 or new_x >= m or new_y < 0 or new_y >= n:
                        continue
                    if grid[new_x][new_y] != 1:
                        continue

                    # mark and and to next level
                    fresh -= 1
                    grid[new_x][new_y] = 2
                    q.put((new_x, new_y))
            steps += 1

        return steps if fresh == 0 else -1

grid = [[2,1,1],[1,1,0],[0,1,1]]
result = Solution().orangesRotting(grid)
print(result)